urlcdd0c1ab49_emiln_JHouseFireExplorer_tgz-pJ8-dk_brics_automaton_DatatypesJ8/src/dk/brics/automaton/BasicOperations.java

```diff
@@ -0,0 +1,699 @@
+	 * already deterministic.
+	 * <p>
+	 * Complexity: quadratic in number of states (if already deterministic).
+	 */
+	static public Automaton minus(Automaton a1, Automaton a2) {
+		if (a1.isEmpty() || a1 == a2)
+			return BasicAutomata.makeEmpty();
+		if (a2.isEmpty())
+			return a1.cloneIfRequired();
+		if (a1.isSingleton()) {
+			if (a2.run(a1.singleton))
+				return BasicAutomata.makeEmpty();
+			else
+				return a1.cloneIfRequired();
+		}
+		return intersection(a1, a2.complement());
+	}
+
+	/**
+	 * Returns an automaton that accepts the intersection of
+	 * the languages of the given automata. 
+	 * Never modifies the input automata languages.
+	 * <p>
+	 * Complexity: quadratic in number of states.
+	 */
+	static public Automaton intersection(Automaton a1, Automaton a2) {
+		if (a1.isSingleton()) {
+			if (a2.run(a1.singleton))
+				return a1.cloneIfRequired();
+			else
+				return BasicAutomata.makeEmpty();
+		}
+		if (a2.isSingleton()) {
+			if (a1.run(a2.singleton))
+				return a2.cloneIfRequired();
+			else
+				return BasicAutomata.makeEmpty();
+		}
+		if (a1 == a2)
+			return a1.cloneIfRequired();
+		Transition[][] transitions1 = Automaton.getSortedTransitions(a1.getStates());
+		Transition[][] transitions2 = Automaton.getSortedTransitions(a2.getStates());
+		Automaton c = new Automaton();
+		LinkedList<StatePair> worklist = new LinkedList<StatePair>();
+		HashMap<StatePair, StatePair> newstates = new HashMap<StatePair, StatePair>();
+		StatePair p = new StatePair(c.initial, a1.initial, a2.initial);
+		worklist.add(p);
+		newstates.put(p, p);
+		while (worklist.size() > 0) {
+			p = worklist.removeFirst();
+			p.s.accept = p.s1.accept && p.s2.accept;
+			Transition[] t1 = transitions1[p.s1.number];
+			Transition[] t2 = transitions2[p.s2.number];
+			for (int n1 = 0, b2 = 0; n1 < t1.length; n1++) {
+				while (b2 < t2.length && t2[b2].max < t1[n1].min)
+					b2++;
+				for (int n2 = b2; n2 < t2.length && t1[n1].max >= t2[n2].min; n2++) 
+					if (t2[n2].max >= t1[n1].min) {
+						StatePair q = new StatePair(t1[n1].to, t2[n2].to);
+						StatePair r = newstates.get(q);
+						if (r == null) {
+							q.s = new State();
+							worklist.add(q);
+							newstates.put(q, q);
+							r = q;
+						}
+						char min = t1[n1].min > t2[n2].min ? t1[n1].min : t2[n2].min;
+						char max = t1[n1].max < t2[n2].max ? t1[n1].max : t2[n2].max;
+						p.s.transitions.add(new Transition(min, max, r.s));
+					}
+			}
+		}
+		c.deterministic = a1.deterministic && a2.deterministic;
+		c.removeDeadTransitions();
+		c.checkMinimizeAlways();
+		return c;
+	}
+		
+	/**
+	 * Returns true if the language of <code>a1</code> is a subset of the
+	 * language of <code>a2</code>. 
+	 * As a side-effect, <code>a2</code> is determinized if not already marked as
+	 * deterministic.
+	 * <p>
+	 * Complexity: quadratic in number of states.
+	 */
+	public static boolean subsetOf(Automaton a1, Automaton a2) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 16. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.